tools/codegen/api/types.py

```diff
@@ -72,7 +72,69 @@ def cpp_type(self, *, strip_ref: bool = False) -> str:
     def name(self) -> ArgName:
         return self.elem.name
 
-CType = Union[BaseCType, OptionalCType, ConstRefCType, MutRefCType]
+@dataclass(frozen=True)
+class ListCType:
+    elem: 'CType'
+
+    def cpp_type(self, *, strip_ref: bool = False) -> str:
+        # Do not pass `strip_ref` recursively.
+        return f'c10::List<{self.elem.cpp_type()}>'
+
+    @property
+    def name(self) -> ArgName:
+        return self.elem.name
```

Comment:
Aha, turning `CType` into a little ADT is shining a brighter light on a structural problem with how "semantic information" (name, currently) is modeled here - it should really be associated with a complete type expression, not an expression's leaf nodes. 

The original composites skated on having exactly one leaf node, and most of the new ones here do too, but ofc the cracks are showing in `TupleCType` - also picture if we do a Map<K, V> type, etc.

It should be easy to refactor (and we should) but I wouldn't necessarily try to jam it into this PR, your call.

BTW, why not convert the originals (`OptionalCType` et al) as well?